Mathlib/Combinatorics/SimpleGraph/Walk.lean

```diff
@@ -995,17 +1023,28 @@ theorem exists_boundary_dart {u v : V} (p : G.Walk u v) (S : Set V) (uS : u ∈
       exact ⟨d, List.Mem.tail _ hd, hcd⟩
     · exact ⟨⟨(x, y), a⟩, List.Mem.head _, uS, h⟩
 
-lemma getVert_tail {u v n} (p : G.Walk u v) (hnp: ¬ p.Nil) :
-    (p.tail hnp).getVert n = p.getVert (n + 1) :=
-  p.notNilRec (fun _ _ ↦ by simp only [tail_cons, cons_getVert_succ]) hnp
-
-@[simp]
-lemma cons_sndOfNotNil (q : G.Walk v w) (hadj : G.Adj u v) :
-    (q.cons hadj).sndOfNotNil not_nil_cons = v := by
-  unfold sndOfNotNil; simp only [notNilRec_cons]
-
-lemma getVert_one (p : G.Walk u v) (hnp : ¬ p.Nil) : p.getVert 1 = p.sndOfNotNil hnp :=
-  p.notNilRec (fun _ _ ↦ by simp only [cons_getVert_succ, getVert_zero, cons_sndOfNotNil]) hnp
+@[simp] lemma getVert_copy  {u v w x: V} (p : G.Walk u v) (i : ℕ) (h : u = w) (h' : v = x) :
+    (p.copy h h').getVert i = p.getVert i := by
+  cases p with
+  | nil =>
+    rw [getVert_of_length_le _ (by simp only [length_nil, Nat.zero_le] : nil.length ≤ _)]
+    rw [getVert_of_length_le _ (by simp only [length_copy, length_nil, Nat.zero_le])]
+    exact h'.symm
+  | cons hadj q =>
+    by_cases hi : i = 0
+    · simp only [hi, getVert_zero, h.symm]
+    · simp only [copy_cons, getVert_cons_succ]
+      rw [getVert_cons _ _ hi]
+      rw [getVert_cons _ _ hi]
+      apply getVert_copy
```

Comment:
```suggestion
  cases p, i with
  | nil, _ =>
    rw [getVert_of_length_le _ (by simp only [length_nil, Nat.zero_le] : nil.length ≤ _)]
    rw [getVert_of_length_le _ (by simp only [length_copy, length_nil, Nat.zero_le])]
    exact h'.symm
  | cons hadj q, 0 => simp only [hi, getVert_zero, h.symm]
  | cons hadj q, i + 1 => simp only [copy_cons, getVert_cons_succ, getVert_copy]
```
or something like that

In fact, since the `cases` is the first thing in the proof, it would be better to even get rid of the tactic mode proof and do pattern matching instead

Reply:
There's something funky going on here that I can't quite put my finger on.

A match outside tactic mode complains about a type mismatch for the nil.
A match inside tactic mode errors with `Case tag 'rhs' not found.`
It seems to have something to do with the copy in the statement, as soon as i remove that there's no problems. 
cases also doesn't seem to support multiple arguments? I tried something with a product, but that didn't seem to work either.

Any pointers?:)

Reply:
@YaelDillies I accidently resolved this conversation, so just a friendly tag to make sure it's doesn't disappear in the void

Reply:
Fixed this by using subst_vars